src/main/java/io/github/thebusybiscuit/cscorelib2/blocks/BlockPosition.java

```diff
@@ -0,0 +1,179 @@
+     *
+     * @return The position of this block.
+     */
+    public long getPosition() {
+        return position;
+    }
+
+    /**
+     * Gets the x for this block.
+     *
+     * @return This blocks x coordinate.
+     */
+    public int getX() {
+        return (int) (this.position >> 38);
+    }
+
+    /**
+     * Gets the y for this block.
+     *
+     * @return This blocks y coordinate.
+     */
+    public int getY() {
+        return (int) (this.position & 0xFFF);
+    }
+
+    /**
+     * Gets the z for this block.
+     *
+     * @return This blocks z coordinate.
+     */
+    public int getZ() {
+        return (int) (this.position << 26 >> 38);
+    }
+
+    /**
+     * Gets the {@link Block} at this position. Note, Bukkit will create a new instance so if you can avoid doing this
+     * then do as it is a bit costly.
+     *
+     * @return The {@link Block} at this location.
+     */
+    public Block getBlock() {
+        return getChunk().getBlock((getX() & 0xF), getY(), (getZ() & 0xF));
+    }
+
+    /**
+     * Gets the {@link Chunk} where this block is located.
+     *
+     * @return This blocks {@link Chunk}.
+     */
+    public Chunk getChunk() {
+        final World world = this.world.get();
+        if (world == null) {
+            throw new IllegalStateException("Cannot get Chunk when the world isn't loaded");
+        }
+
+        return world.getChunkAt(getChunkX(), getChunkZ());
+    }
+
+    /**
+     * Gets the chunks x coordinate for this block.
+     *
+     * @return The blocks chunks x coordinate.
+     */
+    public int getChunkX() {
+        return this.getX() >> 4;
+    }
+
+    /**
+     * Gets the chunks z coordinate for this block.
+     *
+     * @return The blocks chunks z coordinate.
+     */
+    public int getChunkZ() {
+        return this.getZ() >> 4;
+    }
+
+    /**
+     * Transform this BlockPosition into a standard Bukkit {@link Location}.
+     *
+     * @return A Bukkit {@link Location}.
+     */
+    public Location toLocation() {
+        return new Location(this.world.get(), getX(), getY(), getZ());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean equals(Object o) {
+        if (o instanceof BlockPosition) {
+            final BlockPosition pos = (BlockPosition) o;
+            if (pos.world.get() == null) return false;
+            
+            return this.getWorld().getUID().equals(pos.getWorld().getUID())
+                    && this.position == pos.position;
+        }
+        return false;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int hashCode() {
+        final int prime = 31;
+        final World ref = this.world.get();
+    
+        int result = 0;
+        result += prime * (ref == null ? 0 : ref.hashCode());
+        result += prime * Long.hashCode(position);
+        
+        return result;
```

Comment:
May I ask why you don't just use Objects.hash(...) here?

Reply:
I find it quite unnessacry when I can just create my own. Especially if anything was to change or didn't want to be handled in such way.